### ISSUE TITLE
paddle 1.5 训练时loss出Nan

### ISSUE THREAD

**Issue by wzgwzg** [PaddlePaddle/Paddle#18460]
paddle 1.5 训练一个yolov3+分类的multi-task网络时，loss出现Nan的情况。

关键代码：
 self.proposals = self.get_proposals()
            
        axes = [1] 
        starts = [2] 
        ends = [6] 
        self.new_proposals = fluid.layers.slice(self.proposals, axes=axes, starts=starts, ends=ends)
        if self.is_train:
            outs = fluid.layers.generate_proposal_labels(
                    rpn_rois=self.new_proposals,
                    gt_classes=self.gt_reid_label,
                    is_crowd=self.is_crowd_rcnn,
                    gt_boxes=self.gtbox_rcnn,
                    im_info=self.im_info_rcnn,
                    batch_size_per_im=20,
                    fg_fraction=1.0,
                    fg_thresh=0.5,
                    bg_thresh_hi=0.3,
                    bg_thresh_lo=0.0,
                    bbox_reg_weights=[0.1, 0.1, 0.2, 0.2],
                    class_nums=cfg.reid_class_num + 1,
                    use_random=self.use_random)
            
            self.rois = outs[0]
            self.labels_int32 = outs[1]
        if self.is_train:
            pool_rois = self.rois
        else:
            pool_rois = self.new_proposals
        self.roi_align_out = fluid.layers.roi_align(
            input=self.outputs[-1],
            rois=pool_rois,
            pooled_height=7,
            pooled_width=7,
            spatial_scale=1/8.0,
            sampling_ratio=0)
         
        conv_reid_1 = fluid.layers.conv2d(
                input=self.roi_align_out,
                num_filters=128,
                filter_size=1,
                stride=1,
                padding=0,
                act=None,
                param_attr=ParamAttr(
                    learning_rate=10,
                    initializer=fluid.initializer.Normal(0., 0.02),
                    name="conv_reid_1.weights"),
                bias_attr=ParamAttr(
                    learning_rate=10,
                    initializer=fluid.initializer.Constant(0.0),
                    regularizer=L2Decay(0.),
                    name="conv_reid_1.bias"))
         
        relu_reid_1 = fluid.layers.relu(conv_reid_1, name="relu_reid_1")
        conv_reid_2 = fluid.layers.conv2d(
                input=relu_reid_1,
                num_filters=128,
                filter_size=1,
                stride=1,
                padding=0,
                act=None,
                param_attr=ParamAttr(
                    learning_rate=10,
                    initializer=fluid.initializer.Normal(0., 0.02),
                    name="conv_reid_2.weights"),
                bias_attr=ParamAttr(
                    learning_rate=10,
                    initializer=fluid.initializer.Constant(0.0),
                    regularizer=L2Decay(0.),
                    name="conv_reid_2.bias"))
            
        relu_reid_2 = fluid.layers.relu(conv_reid_2, name="relu_reid_2")
         
        conv_reid_3 = fluid.layers.conv2d(
                input=relu_reid_2,
                num_filters=128,
                filter_size=1,
                stride=1,
                padding=0,
                act=None,
                param_attr=ParamAttr(
                    learning_rate=10,
                    initializer=fluid.initializer.Normal(0., 0.02),
                    name="conv_reid_3.weights"),
                bias_attr=ParamAttr(
                    learning_rate=10,
                    initializer=fluid.initializer.Constant(0.0),
                    regularizer=L2Decay(0.),
                    name="conv_reid_3.bias"))
           self.reid_out = fluid.layers.pool2d(conv_reid_3, pool_type='avg', pool_size=7, name='reid_pool')
        self.cls_score = fluid.layers.fc(input=self.reid_out,
                                         size=cfg.reid_class_num + 1,
                                         act=None,
                                         name='reid_cls_score',
                                         param_attr=ParamAttr(
                                            name='reid_cls_score_w',
                                            learning_rate=10, 
                                            initializer=Normal(
                                                 loc=0.0, scale=0.001)),
                                         bias_attr=ParamAttr(
                                             name='reid_cls_score_b',
                                             learning_rate=10,
                                             initializer=fluid.initializer.Constant(0.0),
                                             regularizer=L2Decay(0.)))
        if self.is_train:
            reid_loss = self.reid_loss() * 100
            self.losses.append(reid_loss)
        else:
            self.get_pred()
        
 def get_proposals(self, proposal_thresh=0.005):
         yolo_boxes = fluid.layers.concat(self.boxes, axis=1)
         yolo_scores = fluid.layers.concat(self.scores, axis=2)
         if self.is_train:
             return fluid.layers.multiclass_nms(
                    bboxes=yolo_boxes,
                    scores=yolo_scores,
                    score_threshold=0.0,
                    nms_top_k=800,
                    keep_top_k=200,
                    nms_threshold=0.7,
                    background_label=-1,
                    name="multiclass_nms")
         else:
             return fluid.layers.multiclass_nms(
                    bboxes=yolo_boxes,
                    scores=yolo_scores,
                    score_threshold=0.1,
                    nms_top_k=400,
                    keep_top_k=150,
                    nms_threshold=0.45,
                    background_label=-1,
                    name="multiclass_nms")
        
        
    def reid_loss(self):
        labels_int64 = fluid.layers.cast(x=self.labels_int32, dtype='int64')
        labels_int64.stop_gradient = True
        loss_cls = fluid.layers.softmax_with_cross_entropy(
            logits=self.cls_score,
            label=labels_int64,
            numeric_stable_mode=True,
            ignore_index=0)
        
        loss_cls = fluid.layers.reduce_mean(loss_cls)
        return loss_cls
        
        
    def loss(self):
        return sum(self.losses), sum(self.losses[:-1]), self.losses[-1] 

训练日志：
Iter 0, lr 0.000000, average loss 8.5801, smooth_loss 171.6026, cur_loss 171.6026, yolo_loss 8.5679, reid_loss 0.0122, average time: 6.69479370117e-05
Iter 20, lr 0.000013, average loss 188.0333, smooth_loss 187.2509, cur_loss 204.8698, yolo_loss 187.8873, reid_loss 0.1460, average time: 4.95043145418
Iter 40, lr 0.000027, average loss 197.7558, smooth_loss 192.3752, cur_loss 174.9102, yolo_loss 197.6053, reid_loss 0.1505, average time: 1.99567084312
Iter 60, lr 0.000040, average loss 165.5284, smooth_loss 183.5730, cur_loss 145.6012, yolo_loss 165.3861, reid_loss 0.1423, average time: 2.01746885777
Iter 80, lr 0.000053, average loss 161.2878, smooth_loss 178.0705, cur_loss 127.9836, yolo_loss 161.1663, reid_loss 0.1215, average time: 1.96969819069
Iter 100, lr 0.000067, average loss 142.7933, smooth_loss 171.0849, cur_loss 130.5658, yolo_loss 142.6354, reid_loss 0.1578, average time: 1.86034370661
Iter 120, lr 0.000080, average loss 137.6201, smooth_loss 165.5535, cur_loss 142.1981, yolo_loss 137.3955, reid_loss 0.2246, average time: 2.26731359959
Iter 140, lr 0.000093, average loss 139.1105, smooth_loss 161.8027, cur_loss 177.0392, yolo_loss 138.8696, reid_loss 0.2409, average time: 1.7978085041
Iter 160, lr 0.000107, average loss 125.0568, smooth_loss 157.2380, cur_loss 104.0745, yolo_loss 124.8585, reid_loss 0.1983, average time: 2.14733464718
Iter 180, lr 0.000120, average loss 120.6611, smooth_loss 153.1964, cur_loss 114.3480, yolo_loss 120.4586, reid_loss 0.2025, average time: 1.86390209198
Iter 200, lr 0.000133, average loss 125.3435, smooth_loss 150.4250, cur_loss 96.7701, yolo_loss 125.1871, reid_loss 0.1564, average time: 2.14077150822
Iter 220, lr 0.000147, average loss 124.2083, smooth_loss 148.0524, cur_loss 103.6380, yolo_loss 124.0151, reid_loss 0.1932, average time: 1.99593420029
Iter 240, lr 0.000160, average loss 121.8349, smooth_loss 145.8767, cur_loss 164.4178, yolo_loss 121.6617, reid_loss 0.1732, average time: 1.96769964695
Iter 260, lr 0.000173, average loss 125.6951, smooth_loss 144.3302, cur_loss 136.9492, yolo_loss 125.5389, reid_loss 0.1562, average time: 2.02929550409
Iter 280, lr 0.000187, average loss nan, smooth_loss nan, cur_loss nan, yolo_loss nan, reid_loss nan, average time: 2.10160044432
Iter 300, lr 0.000200, average loss nan, smooth_loss nan, cur_loss nan, yolo_loss nan, reid_loss nan, average time: 1.84718315601
Iter 320, lr 0.000213, average loss nan, smooth_loss nan, cur_loss nan, yolo_loss nan, reid_loss nan, average time: 2.12868974209
Iter 340, lr 0.000227, average loss nan, smooth_loss nan, cur_loss nan, yolo_loss nan, reid_loss nan, average time: 1.94317824841
Iter 360, lr 0.000240, average loss nan, smooth_loss nan, cur_loss nan, yolo_loss nan, reid_loss nan, average time: 1.80638080835

**Comment by heavengate** [comment]
可尝试一下方式：
1. 增加batch size或降低学习率
2. 使用gradient clip， 参考http://www.paddlepaddle.org.cn/documentation/docs/zh/1.5/api_cn/clip_cn.html#paddle.fluid.clip.GradientClipByGlobalNorm

**Comment by heavengate** [comment]
1. reid_loss = self.reid_loss() * 100 中乘以100对梯度影响很大，建议去掉
2. def get_proposals(self, proposal_thresh=0.005): 中multiclass_nms是没有梯度的，会导致整个yolo网络没有训练，而且输出是一个变长的lod tensor，会导致后续数据错乱

**Comment by heavengate** [comment]
1. multi nms的输出是lod tensor，你后面这些conv2d这些操作是针对普通tensor的，这个数据过去应该不能正确解析
2.multi nms在检测到bbox的时候输出是[N, 6], 6那一维是class_id, score, xmin, ymin, xmax, ymax, 直接过后面的conv2d会有问题吧
3. multi nms在图像上检测不到bbox的时候shape是[1, 1]，值是0， 这样的值后面没做特殊处理的话训练可能会引入问题
